examples/less_than.rs

```diff
@@ -0,0 +1,234 @@
+use bellpepper_core::{
+  boolean::AllocatedBit, num::AllocatedNum, Circuit, ConstraintSystem, LinearCombination,
+  SynthesisError,
+};
+use ff::{PrimeField, PrimeFieldBits};
+use pasta_curves::Fq;
+use spartan2::{
+  errors::SpartanError,
+  traits::{snark::RelaxedR1CSSNARKTrait, Group},
+  SNARK,
+};
+
+fn num_to_bits_le_bounded<F: PrimeField + PrimeFieldBits, CS: ConstraintSystem<F>>(
+  cs: &mut CS,
+  n: AllocatedNum<F>,
+  num_bits: u8,
+) -> Result<Vec<AllocatedBit>, SynthesisError> {
+  let opt_bits = match n.get_value() {
+    Some(v) => v
+      .to_le_bits()
+      .into_iter()
+      .take(num_bits as usize)
+      .map(Some)
+      .collect::<Vec<Option<bool>>>(),
+    None => vec![None; num_bits as usize],
+  };
+
+  // Add one witness per input bit in little-endian bit order
+  let bits_circuit = opt_bits.into_iter()
+    .enumerate()
+    // AllocateBit enforces the value to be 0 or 1 at the constraint level
+    .map(|(i, b)| AllocatedBit::alloc(cs.namespace(|| format!("b_{}", i)), b).unwrap())
+    .collect::<Vec<AllocatedBit>>();
+
+  let mut weighted_sum_lc = LinearCombination::zero();
+  let mut pow2 = F::ONE;
+
+  for bit in bits_circuit.iter() {
+    weighted_sum_lc = weighted_sum_lc + (pow2, bit.get_variable());
+    pow2 = pow2.double();
+  }
+
+  cs.enforce(
+    || "bit decomposition check",
+    |lc| lc + &weighted_sum_lc,
+    |lc| lc + CS::one(),
+    |lc| lc + n.get_variable(),
+  );
+
+  Ok(bits_circuit)
+}
+
+fn get_msb_index<F: PrimeField + PrimeFieldBits>(n: F) -> u8 {
+  n.to_le_bits()
+    .into_iter()
+    .enumerate()
+    .rev()
+    .find(|(_, b)| *b)
+    .unwrap()
+    .0 as u8
+}
+
+// Range check: constrains input < `bound`. The bound must fit into
+// `num_bits` bits (this is asserted in the circuit constructor).
+// Important: it must be checked elsewhere (in an overarching circuit) that the
+// input fits into `num_bits` bits - this is NOT constrained by this circuit
+// in order to avoid duplications (hence "unsafe"). Cf. LessThanCircuitSafe for
+// a safe version.
+#[derive(Clone, Debug)]
+struct LessThanCircuitUnsafe<F: PrimeField> {
+  bound: F, // Will be a constant in the constraits, not a variable
```

Comment:
I think that a more natural representation of `bound` is as unsigned integer, rather than a field element. For most cases a `u8` would suffice, but with `u16` you're definitely good.

Reply:
Yes, in fact originally both the input and the bound were `u64`s (cf. [here](https://github.com/HungryCatsStudio/Spartan2/blob/b7bf232d65fdb246af7f71af984d82e7c28031f6/examples/less_than.rs#L107-L108)). Then we needed to change the input to `F` because we realised only then can one (in general) show examples where the safe detects cheating which the unsafe version doesn't. And, since the input became an `F`, it made sense to us to have the bound as an `F` as well. This change of the bound type in fact came at the cost of programming of [a small extra function](https://github.com/HungryCatsStudio/Spartan2/blob/antonio-cesar/less-than/examples/less_than.rs#L53) by @Cesar199999 because an original constructor check could no longer be applied (original check [here](https://github.com/HungryCatsStudio/Spartan2/blob/b7bf232d65fdb246af7f71af984d82e7c28031f6/examples/less_than.rs#L66), new check [here](https://github.com/HungryCatsStudio/Spartan2/blob/antonio-cesar/less-than/examples/less_than.rs#L78)).

I think having the bound as `F` is more general: one can always use `F::from` if one has an unsigned integer bound, but the converse cannot be done if the bound is passed to the constructor as a `u_something` and one has an `F`. Furthermore, most fields we use are larger than even `u128`, so with an `F` one can express more bounds. Note that this doesn't cause extra constraints, since the number of constraints is dictated by the user-defined variable `num_bits`. This is an example for the library, so making it more general than our specific needs doesn't seem pointless.

Having said all of that, we can totally change it to a `u_something` if you prefer. I'd say `u16` is unnecessarily small, maybe `u64` (as originally) better? That's also the type `F::from` wants. If we make this change, we could do away with `get_msb_index` (going back to the original check) and add the check that the new unsigned bound fits into the field.

Reply:
Oh I thought this is meant to represent the bound on the number of bits, rather than the bound itself, hence I proposed `u16`.

Reply:
Kind of like the `shift_bound` in the r1cs exercises: https://github.com/rdi-berkeley/zkp-mooc-lab/blob/main/circuits/float_add.circom#L256

Reply:
Not quite, so: these chips implement the check `input < bound`, where the LHS is a witness and the RHS is a constant. So, in [here](https://github.com/HungryCatsStudio/Spartan2/blob/eb2843e26ff653d074ecaef5b1a47876e832f964/examples/less_than.rs#L124-L126), `input` and `bound` are as I just described, and `num_bits` represents what input size should be supported. For instance, if `num_bits = 32`, then inputs between 0 and 2^32 - 1 are supported. This is of course different from the number of bits of the bound: in the previous example, one could have e.g. `bound = 2024`, which is 11 bits long (but we might want to compare inputs between 0 and 2^32 - 1).

Reply:
I have just updated the documentation, which described the circuit as a range check, since that might be misleading (i.p. it sounds like the bound is of the form `2^n`, which might have confused you @mmagician)

Reply:
Ah okey, yes makes sense.